site/app/controllers/AuthenticationController.php

```diff
@@ -308,4 +310,274 @@ public function userSelection() {
+            $is_given_name = (strtolower($id_given_name) === substr(strtolower($given_name), 0, $name_requirements['given_name']));
+            $is_family_name = (strtolower($id_family_name) === substr(strtolower($family_name), 0, $name_requirements['family_name']));
+            if ($is_family_name && $is_given_name) {
+                return true;
+            }
+            return false;
+        }
+        elseif ($requirements['require_email'] === true) {
+            if ($requirements['email_requirements']['whole_email']) {
+                return $user_id === $email;
+            }
+            elseif ($requirements['email_requirements']['whole_prefix']) {
+                $split_email = explode('@', $email);
+                $email_extension = array_pop($split_email);
+                return $user_id === implode('', $split_email);
+            }
+            else {
+                return substr($user_id, 0, $requirements['email_requirements']['prefix_count']) === substr($email, 0, $requirements['email_requirements']['prefix_count']);
+            }
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns true if the password is greater than or equal to 12 characters, and has the required characters
+     */
+    public function isGoodPassword(string $password): bool {
+        return strlen($password) >= 12 && $this->checkChars($password);
+    }
+
+    /**
+     * Checks if the email extension is in the accepted emails part of the Submitty config file
+     */
+    public function isAcceptedEmail(string $email): bool {
+        $emails = $this->core->getConfig()->getAcceptedEmails();
+        // Check if the file was read successfully
+        try {
+            $split_email = explode('@', $email);
+            $email_extension = $split_email[count($split_email) - 1];
+        }
+        catch (\Error $error) {
+            return false;
+        }
+        return in_array($email_extension, array_keys($emails), true);
+    }
+
+    /**
+     * @return array<mixed>
+     */
+    public function generateVerificationCode(): array {
+        $code = $this->core->getConfig()->isDebug() ? '00000000' : Utils::generateRandomString();
+        $timestamp = time() + 60 * 15; // 15 minutes from now, may eventually set this as a configurable value.
+        return ['code' => strval($code), 'exp' => $timestamp];
+    }
+
+    public function sendVerificationEmail(string $email, string $verification_code, string $user_id): void {
+        $subject = "Submitty Email Verification";
+        $url = $this->core->getConfig()->getBaseUrl() . 'authentication/verify_email?verification_code=' . $verification_code;
+        $body = <<<EMAIL
+                Welcome to Submitty! We are excited to have you on board. To complete your account setup, either enter this verification code, or click the link below.
+
+                Verification Code: $verification_code
+
+                Verification Link: $url
```

Comment:
Is there a reason why we need a link as well as a code?  Perhaps just a code would be sufficient?

Reply:
When talking with Barb about this a few months ago, she said it would be good to have both. My view is if you accidentally close the page that you enter the code into, you can just use the link. I am not partial to either way, but I think more in person discussion with more people would be beneficial.

Reply:
That seems like a reasonable justification to me.  Thanks for the clarification!